### PR TITLE
feat(scoop-bucket): List more detailed information for buckets

### PR DESCRIPTION
**Pull request by HUMORCE**
#### Description
show detailed list for buckets.

#### Motivation and Context
this is a missing part of `scoop-bucket`, IMO

#### How Has This Been Tested?
**before**
```
> scoop bucket list
1ocal
extras
java
main
nuke
tests
versions
```

**after**
```
> scoop bucket list

Name     Source                                     Updated             Manifests
----     ------                                     -------             ---------
1ocal    ~\scoop\buckets\1ocal                      2022-02-02 17:58:02         3
extras   git@github.com:ScoopInstaller/Extras.git   2022-02-03 00:43:11      1380
java     git@github.com:ScoopInstaller/Java.git     2022-02-02 09:58:43       215
main     git@github.com:ScoopInstaller/Main.git     2022-02-02 16:29:48       971
nuke     git@github.com:HUMORCE/nuke.git            2022-02-02 12:09:58       107
tests    git@github.com:ScoopInstaller/Tests.git    2022-01-30 01:32:15        51
versions git@github.com:ScoopInstaller/Versions.git 2022-02-03 00:54:07       276
```

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

File: libexec/scoop-bucket.ps1 [hunk 1]
```diff
@@ -19,19 +19,43 @@
...
+        }
+
+        $buckets += New-Object PSObject -Property @{
+            Name    = $bucket
+            Source  = $source
+            Updated = $updated
+        }
+    }
+
+    return $buckets | Select-Object Name, Source, Updated | Format-Table -AutoSize -Wrap
```
Comment (rashil2000) [inline libexec/scoop-bucket.ps1:52]:
```suggestion
    return $buckets | Select-Object Name, Source, Updated
```

PowerShell does it automatically, so not required.

**Comment by rashil2000** [comment]
Would it be cool to also show the number of apps in each bucket? I think it's useful.

File: libexec/scoop-bucket.ps1 [hunk 2]
```diff
@@ -19,19 +19,48 @@
...
+            Measure-Object | Select-Object -ExpandProperty Count
+        )
+        $updated = 'N/A'
+        if (Test-Path (Join-Path $source '.git')) {
+            $updated = git_cmd -C "`"$source`"" log --format='%ch' -n 1
+            $source = git_cmd -C "`"$source`"" config remote.origin.url
+        }
+        if ($source.StartsWith((Get-PSProvider 'FileSystem').Home)) {
+            $source = $source.Replace((Get-PSProvider 'FileSystem').Home, '~')
+        }
```
Comment (rashil2000) [inline libexec/scoop-bucket.ps1:47]: Use the `friendly_path` function for this: https://github.com/ScoopInstaller/Scoop/blob/develop/lib/core.ps1#L381

File: libexec/scoop-bucket.ps1 [hunk 3]
```diff
@@ -19,19 +19,48 @@
...
+        $source = Find-BucketDirectory $bucket -Root
+        $manifests = (
+            Get-ChildItem "$source\bucket" -Force -Recurse -ErrorAction SilentlyContinue |
+            Measure-Object | Select-Object -ExpandProperty Count
+        )
+        $updated = 'N/A'
+        if (Test-Path (Join-Path $source '.git')) {
+            $updated = git_cmd -C "`"$source`"" log --format='%ch' -n 1
+            $source = git_cmd -C "`"$source`"" config remote.origin.url
+        }
```
Comment (rashil2000) [inline libexec/scoop-bucket.ps1:44]: We can use `(Get-Item $source).LastWriteTime` for folders that are not git repos?

Comment (rashil2000) [reply libexec/scoop-bucket.ps1]: Yeah, I think that's alright. Better than no time at all.

Comment (HUMORCE) [reply libexec/scoop-bucket.ps1]: or both with `yyyy-MM-dd HH:mm:ss`?

```
Name     Manifests Source                                     Updated
----     --------- ------                                     -------
1ocal            3 ~\scoop\buckets\1ocal                      2022-02-02 17:48:10
extras        1380 git@github.com:ScoopInstaller/Extras.git   2022-02-03 00:43:11
```

Comment (rashil2000) [reply libexec/scoop-bucket.ps1]: That works too 👍

**Comment by rashil2000** [comment]
Hmm, why is friendly_path not working for me 🤔

image

image

File: libexec/scoop-bucket.ps1 [hunk 4]
```diff
@@ -45,16 +45,14 @@ function list_buckets {
...
         $buckets += New-Object PSObject -Property @{
             Name      = $bucket
             Source    = $source
             Updated   = $updated
             Manifests = $manifests
         }
     }
-
-    return $buckets | Select-Object Name, Source, Updated, Manifests
+    return ($buckets | Select-Object Name, Source, Updated, Manifests | Out-String).Trim()
```
Comment (rashil2000) [inline libexec/scoop-bucket.ps1:55]: Why out-string? It is better to return PSObject, for manipulating pipeline (i.e. `scoop bucket list | select manifests`)

Comment (HUMORCE) [reply libexec/scoop-bucket.ps1]: wanted to remove blank lines

Comment (rashil2000) [reply libexec/scoop-bucket.ps1]: Oh. I'm afraid there's no way to do that. PowerShell adds newlines when printing objects.

**Comment by HUMORCE** [comment]
> Hmm, why is friendly_path not working for me 🤔

because bucket 'okay' is git repo.

```
> scoop bucket add cewnqj D:\Downloads\     
Checking repo... 'D:\Downloads\' doesn't look like a valid git repository

Error given:
fatal: 'D:\Downloads\' does not appear to be a git repository fatal: Could not read from remote repository. System.Management.Automation.RemoteException Please make sure you have the correct access rights and the repository exists.
```

File: libexec/scoop-bucket.ps1 (hunk 4 above)
Comment (HUMORCE) [reply libexec/scoop-bucket.ps1]: 3 blank lines after output in PS 5.1
2 blank lines after output in PS 7.2

**Comment by rashil2000** [comment]
Hmm, local buckets aren't fully supported as of now. Can be solved later. 

Other than that, this PR looks great. One minor nit: I think this should be `feat(scoop-bucket)` instead of `refactor(scoop-bucket)`, because it doesn't refactor anything but adds new features.

**Comment by HUMORCE** [comment]
it is kind like replacing the entire hardware of a computer step by step, but it is still actually a computer at end.

i dont know how to define them in this case, but i take your advice.

**Comment by niheaven** [comment]
> it is kind like replacing the entire hardware of a computer step by step, but it is still actually a computer at end.
> 
> i dont know how to define them in this case, but i take your advice.

`refactor` is refactoring some code, without change its functionality and UI (maybe a little tweak), while `feature` means add new functionality to some existing function or UI. So yes, `feature` is more precise.

PS, some changes under `refactor` in `CHANGELOG.md` are just `fix` IMO :)

**Pull request merged**